waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,512 @@
+
+    return ok((elements, cursor))
+
+  return ok(callback)
+
+proc createTransferCallback(
+    self: WakuSync, wakuArchive: WakuArchive, wakuStoreClient: WakuStoreClient
+): Result[TransferCallback, string] =
+  if wakuArchive.isNil():
+    return err("waku archive unavailable")
+
+  if wakuStoreClient.isNil():
+    return err("waku store client unavailable")
+
+  let callback: TransferCallback = proc(
+      hashes: seq[WakuMessageHash], peerId: PeerId
+  ): Future[Result[void, string]] {.async: (raises: []), closure.} =
+    var query = StoreQueryRequest()
+    query.includeData = true
+    query.messageHashes = hashes
+    query.paginationLimit = some(uint64(100))
+
+    while true:
+      let catchable = catch:
+        await wakuStoreClient.query(query, peerId)
+
+      if catchable.isErr():
+        return err("store client error: " & catchable.error.msg)
+
+      let res = catchable.get()
+      let response = res.valueOr:
+        return err("store client error: " & $error)
+
+      query.paginationCursor = response.paginationCursor
+
+      for kv in response.messages:
+        let handleRes = catch:
+          await wakuArchive.handleMessage(kv.pubsubTopic.get(), kv.message.get())
+
+        if handleRes.isErr():
+          error "message transfer failed", error = handleRes.error.msg
+          # Messages can be synced next time since they are not added to storage yet.
+          continue
+
+        self.messageIngress(kv.pubsubTopic.get(), kv.messageHash, kv.message.get())
+
+      if query.paginationCursor.isNone():
+        break
+
+    return ok()
+
+  return ok(callback)
+
+proc initFillStorage(
+    self: WakuSync, wakuArchive: WakuArchive
+): Future[Result[void, string]] {.async.} =
+  if wakuArchive.isNil():
+    return err("waku archive unavailable")
+
+  let endTime = getNowInNanosecondTime()
+  let starTime = endTime - self.syncRange.nanos
+
+  var query = ArchiveQuery(
+    includeData: true,
+    cursor: none(ArchiveCursor),
+    startTime: some(starTime),
+    endTime: some(endTime),
+    pageSize: 100,
+  )
+
+  while true:
+    let response = (await wakuArchive.findMessages(query)).valueOr:
+      return err($error)
+
+    for i in 0 ..< response.hashes.len:
+      let hash = response.hashes[i]
+      let topic = response.topics[i]
+      let msg = response.messages[i]
+
+      self.messageIngress(topic, hash, msg)
+
+    if response.cursor.isNone():
+      break
+
+    query.cursor = response.cursor
+
+  return ok()
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncRange: timer.Duration = DefaultSyncRange,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    wakuArchive: WakuArchive,
+    wakuStoreClient: WakuStoreClient,
+    pruneCallback: Option[PruneCallback] = none(PruneCallback),
+    transferCallback: Option[TransferCallback] = none(TransferCallback),
```

Comment:
The `transferCallback` is always unset or set to None. This parameter seems to be not needed

Reply:
For tests we can mock the callbacks but for normal operation we just pass archive and store client so that the callback are created automatically.